src/Algebra/Graph/Undirected.hs

```diff
@@ -0,0 +1,907 @@
+edges in the graph, and /s/ will denote the /size/ of the corresponding
+'Graph' expression. For example, if @g@ is an Undirected 'Graph' then /n/, /m/ and /s/ can
+be computed as follows:
+
+@n == 'vertexCount' g
+m == 'edgeCount' g
+s == 'size' g@
+
+Note that 'size' counts all leaves of the expression:
+
+@'vertexCount' 'empty'           == 0
+'size'        'empty'           == 1
+'vertexCount' ('vertex' x)      == 1
+'size'        ('vertex' x)      == 1
+'vertexCount' ('empty' + 'empty') == 0
+'size'        ('empty' + 'empty') == 2@
+
+Converting an Undirected 'Graph' to the corresponding 'AM.AdjacencyMap' takes /O(s + m * log(m))/
+time and /O(s + m)/ memory. This is also the complexity of the graph equality
+test, because it is currently implemented by converting graph expressions to
+canonical representations based on adjacency maps.
+
+The total order on graphs is defined using /size-lexicographic/ comparison:
+
+* Compare the number of vertices. In case of a tie, continue.
+* Compare the sets of vertices. In case of a tie, continue.
+* Compare the number of edges. In case of a tie, continue.
+* Compare the sets of edges.
+
+Here are a few examples:
+
+@'vertex' 1 < 'vertex' 2
+'vertex' 3 < 'edge' 1 2
+'vertex' 1 < 'edge' 1 1
+'edge' 1 1 < 'edge' 1 2
+'edge' 1 2 < 'edge' 1 1 + 'edge' 2 2
+'edge' 1 2 < 'edge' 1 3@
+'edge' 1 2 == 'edge' 2 1
+
+Note that the resulting order refines the 'isSubgraphOf' relation and is
+compatible with 'overlay' and 'connect' operations:
+
+@'isSubgraphOf' x y ==> x <= y@
+
+@'empty' <= x
+x     <= x + y
+x + y <= x * y@
+-}
+newtype Graph a = UndirectedGraph { graph :: G.Graph a }
+             deriving (Generic, NFData)
+
+instance (Show a, Ord a) => Show (Graph a) where
+  show = show . toSymmetricRelation
+
+{- See Note [Functions for rewrite rules] in 'Algebra.Graph' -}
+
+instance Functor Graph where
+    fmap = fmapR
+
+fmapR :: (a -> b) -> Graph a -> Graph b
+fmapR f g = bindR g (vertex . f)
+{-# INLINE fmapR #-}
+
+-- | __Note:__ this does not satisfy the usual ring laws; see 'Graph' for more
+-- details.
+instance Num a => Num (Graph a) where
+    fromInteger = vertex . fromInteger
+    (+)         = overlay
+    (*)         = connect
+    signum      = const empty
+    abs         = id
+    negate      = id
+
+instance Ord a => Eq (Graph a) where
+    (==) = eqR
+
+instance Ord a => Ord (Graph a) where
+    compare = ordR
+
+-- TODO: Find a more efficient equality check.
+-- TODO: Implement toUndirectedAdjacencyMap.
+-- Check if two graphs are equal by converting them to their adjacency maps.
+eqR :: Ord a => Graph a -> Graph a -> Bool
+eqR x y = toSymmetricRelation x == toSymmetricRelation y
+{-# NOINLINE [1] eqR #-}
+{-# RULES "eqR/Int" eqR = eqIntR #-}
+
+-- Like 'eqR' but specialised for graphs with vertices of type 'Int'.
+eqIntR :: Graph Int -> Graph Int -> Bool
+eqIntR x y = toSymmetricRelation x == toSymmetricRelation y
```

Comment:
At the moment this function is not specialised at all, since we don't have have `Relation` specialised to vertices of type `Int`. Let's keep the code as is, but add a `TODO`, saying that this needs to be fixed in future.

Reply:
Yes, I left it there and meant to leave a note saying that it wasn't really specialized, but forgot! I will fix that!